api/src/main/java/org/motechproject/nms/api/web/UserController.java

```diff
@@ -31,47 +39,123 @@
 
     @Autowired
     private LanguageService languageService;
+
     @Autowired
     private KilkariService kilkariService;
 
+    @Autowired
+    private FrontLineWorkerService frontLineWorkerService;
+
+    @Autowired
+    private ServiceUsageService serviceUsageService;
+
+    @Autowired
+    private ServiceUsageCapService serviceUsageCapService;
+
     @RequestMapping("/{serviceName}/user")
     @ResponseBody
     public ResponseUser user(@PathVariable String serviceName, @RequestParam String callingNumber,
                              @RequestParam String operator, @RequestParam String circle, @RequestParam String callId)
             throws NotFoundException {
-        List<Language> languages = languageService.getCircleLanguages(circle);
         StringBuilder failureReasons = validate(callingNumber, operator, circle, callId);
+        if (failureReasons.length() > 0) {
+            throw new IllegalArgumentException(failureReasons.toString());
+        }
 
         ResponseUser user = null;
 
-        if (MOBILE_ACADEMY.equals(serviceName)) {
-            user = new MobileAcademyUser();
-        } else if (MOBILE_KUNJI.equals(serviceName)) {
-            user = new MobileKunjiUser();
-        } else if (KILKARI.equals(serviceName)) {
-            user = new KilkariResponseUser();
-            Subscriber subscriber = kilkariService.getSubscriber(callingNumber);
-            if (subscriber == null) {
-                throw new NotFoundException(String.format(NOT_FOUND, "callingNumber"));
-            }
-            Set<Subscription> subscriptions = subscriber.getSubscriptions();
-            Set<String> packs = new HashSet<>();
-            for (Subscription subscription : subscriptions) {
-                packs.add(subscription.getSubscriptionPack().getName());
-            }
-            ((KilkariResponseUser) user).setSubscriptionPackList(packs);
-        } else {
+        /*
+        Make sure the url the user hit corresponds to a service we are expecting
+         */
+        if (!(MOBILE_ACADEMY.equals(serviceName) || MOBILE_KUNJI.equals(serviceName)) && !(KILKARI.equals(serviceName))) {
             failureReasons.append(String.format(INVALID, "serviceName"));
         }
 
+        /*
+        Handle the FLW services
+         */
+        if (MOBILE_ACADEMY.equals(serviceName) || MOBILE_KUNJI.equals(serviceName)) {
+            user = getFrontLineWorkerResponseUser(serviceName, callingNumber);
+        }
+
+        /*
+        Kilkari in the house!
+         */
+        if (KILKARI.equals(serviceName)) {
+            user = getKilkariResponseUser(callingNumber);
+        }
+
```

Comment:
I think it is easier to read this way.  Doing it with if -> else if -> else
makes it hard to understand what case falls to the else branch.

On Mon, Apr 20, 2015 at 10:05 AM, Frank Huster <notifications@github.com>
wrote:

> In api/src/main/java/org/motechproject/nms/api/web/UserController.java
> <https://github.com/koshalt/mim/pull/46#discussion_r28707570>:
>
> >              failureReasons.append(String.format(INVALID, "serviceName"));
> >          }
> >
> > +        /*
> > +        Handle the FLW services
> > +         */
> > +        if (MOBILE_ACADEMY.equals(serviceName) || MOBILE_KUNJI.equals(serviceName)) {
> > +            user = getFrontLineWorkerResponseUser(serviceName, callingNumber);
> > +        }
> > +
> > +        /*
> > +        Kilkari in the house!
> > +         */
> > +        if (KILKARI.equals(serviceName)) {
> > +            user = getKilkariResponseUser(callingNumber);
> > +        }
> > +
>
> Why not:
> if (MA || MK) {
> ...
> } else if (KK) {
> ...
> } else {
> invalid service name
> }
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/koshalt/mim/pull/46/files#r28707570>.
>